workers/src/main/scala/workers/Cache.scala

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2016-2022 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package workers
+
+import boopickle.DefaultBasic.*
+import boopickle.Pickler
+import cats.effect.kernel.Async
+import cats.effect.kernel.Sync
+import cats.syntax.all.*
+import explore.model.boopickle.Boopickle.*
+import japgolly.scalajs.react.callback.*
+import japgolly.webapputil.indexeddb.IndexedDb
+import japgolly.webapputil.indexeddb.IndexedDb.Database
+import japgolly.webapputil.indexeddb.IndexedDbKey
+import japgolly.webapputil.indexeddb.KeyCodec
+import japgolly.webapputil.indexeddb.ObjectStoreDef
+import japgolly.webapputil.indexeddb.ValueCodec
+import org.scalajs.dom
+import org.scalajs.dom.CacheStorage
+import org.scalajs.dom.IDBFactory
+import org.scalajs.dom.{Cache => JsCache}
+
+import java.time.Instant
+import java.util.Base64
+import scala.scalajs.js.typedarray.Int8Array
+
+import scalajs.js
+import scalajs.js.JSConverters.*
+
+/**
+ * Cacheable computation.
+ */
+case class Cacheable[F[_], I, O](name: String, version: Int, invoke: I => F[O])
+
+/**
+ * Generic cache interface.
+ */
+sealed trait Cache[F[_]]:
+  def eval[I: Pickler, O: Pickler](computation: Cacheable[F, I, O]): I => F[O]
+
+/**
+ * `NoCache` is used when, for some reason, a cache could not be initialized.
+ */
+case class NoCache[F[_]]() extends Cache[F]:
+  def eval[I: Pickler, O: Pickler](computation: Cacheable[F, I, O]): I => F[O] =
+    computation.invoke
+
+/**
+ * Cache backed up by an Indexed DB store.
+ */
+case class IDBCache[F[_]](
+  cacheDB: IndexedDb.Database,
+  store:   ObjectStoreDef.Sync[Pickled, Pickled]
+)(using
+  F:       Async[F]
+) extends Cache[F]:
+  override def eval[I: Pickler, O: Pickler](computation: Cacheable[F, I, O]): I => F[O] = { input =>
+    val pickledInput: Pickled = Pickled(asBytes((computation.name, computation.version, input)))
+
+    // F.delay(println(s"CACHED EVAL $input")) >>
+    cacheDB
+      .get(store)(pickledInput)
+      .toF
+      // .flatTap(pickled => F.delay(println(pickled.fold("CACHE MISS")(_ => "CACHE HIT"))))
+      .flatMap(
+        _.fold(
+          computation
+            .invoke(input)
+            .flatTap(output => cacheDB.put(store)(pickledInput, Pickled(asBytes(output))).toF)
```

Comment:
We are storing an unpickled timestamp, but in the value instead of the key.

If we were to store it in the key than that would imply either:
- Using 2 stores (one for the keys+timestamps and one for the values); or:
- Using a single store with a compound key, but David's API does not support querying by partial keys, so we would have to switch to the low-level facade for this.

By having the timestamp in the value we can use one store, where the key is the whole input. We use the low-level facade only for eviction, where we sweep the values with a cursor and can then inspect the timestamp without decoding anything.

Reply:
Nice